# Bacterial genome statistics

```{r load_mag_metadata, warning=FALSE, comments="", message=FALSE}
genome_metadata <- read_tsv(file=str_c("data/",release,"_genome.tsv"))
assembly_metadata <- read_tsv(file=str_c("data/",release,"_assembly.tsv"))
host_metadata <- read_tsv(file=str_c("data/",release,"_host.tsv"))

```

```{r ehi_colors, warning=FALSE, comments="", message=FALSE}

phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    select(phylum, colors) %>% 
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

## Genome quantity

```{r genome_quantity_stats, warning=FALSE, comments="", message=FALSE}
genome_metadata %>%
  nrow()
```

## Genome quality

```{r genome_quality_stats, warning=FALSE, comments="", message=FALSE}
genome_metadata %>%
  summarise(completenes_mean=mean(completeness),
            completenes_sd=sd(completeness),
            contamination_mean=mean(contamination),
            contamination_sd=sd(contamination)) %>% 
  tt()
```

```{r genome_quality_plot, warning=FALSE, comments="", message=FALSE}

#Generate quality biplot
genome_biplot <- genome_metadata %>%
  select(c(genome_id,domain,phylum,completeness,contamination,genome_size)) %>%
  ggplot(aes(x=completeness,y=contamination,color=phylum)) +
              geom_point(alpha=0.7, size=1) +
                    ylim(c(10,0)) +
                    scale_color_manual(values=phylum_colors) +
                    labs(y= "Contamination", x = "Completeness") +
                    theme_classic() +
                    theme(legend.position = "none")

#Generate contamination boxplot
genome_contamination <- genome_metadata %>%
            ggplot(aes(y=contamination)) +
                    ylim(c(10,0)) +
                    geom_boxplot(colour = "#999999", fill="#cccccc") +
                    theme_void() +
                    theme(legend.position = "none",
                        axis.title.x = element_blank(),
                        axis.title.y = element_blank(),
                        axis.text.y=element_blank(),
                        axis.ticks.y=element_blank(),
                        axis.text.x=element_blank(),
                        axis.ticks.x=element_blank(),
                        plot.margin = unit(c(0, 0, 0.40, 0),"inches")) #add bottom-margin (top, right, bottom, left)

#Generate completeness boxplot
genome_completeness <- genome_metadata %>%
        ggplot(aes(x=completeness)) +
                xlim(c(50,100)) +
                geom_boxplot(colour = "#999999", fill="#cccccc") +
                theme_void() +
                theme(legend.position = "none",
                    axis.title.x = element_blank(),
                    axis.title.y = element_blank(),
                    axis.text.y=element_blank(),
                    axis.ticks.y=element_blank(),
                    axis.text.x=element_blank(),
                    axis.ticks.x=element_blank(),
                    plot.margin = unit(c(0, 0, 0, 0.50),"inches")) #add left-margin (top, right, bottom, left)

#Render composite figure
grid.arrange(grobs = list(genome_completeness,genome_biplot,genome_contamination),
        layout_matrix = rbind(c(1,1,1,1,1,1,1,1,1,1,1,4),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3)))
```

## Taxonomic statistics

```{r genome_taxonomy_stats, warning=FALSE, comments="", message=FALSE}
#phylum
genome_metadata %>% 
  group_by(phylum) %>%
  summarise(genomes=n(),
            completenes=mean(completeness),
            contamination=mean(contamination),
            size=mean(genome_size)/1000000) %>%
  arrange(-genomes) %>%
  filter(phylum != "p__") %>% 
  tt()

selected_phyla <- genome_metadata %>% 
  group_by(phylum) %>%
  summarise(genomes=n(),
            completenes=mean(completeness),
            contamination=mean(contamination),
            size=mean(genome_size)/1000000) %>%
  arrange(-genomes) %>%
  filter(genomes > 0) %>% 
  filter(phylum != "p__") %>% 
  filter(phylum != "nan") %>% 
  pull(phylum)

#genus
genome_metadata %>% 
  group_by(genus) %>%
  summarise(genomes=n(),
            completenes=mean(completeness),
            contamination=mean(contamination),
            size=mean(genome_size)/1000000) %>%
  arrange(-genomes) %>%
  filter(genus != "g__") %>% 
  tt()

#top 50 genera
genus_top50 <- genome_metadata %>% 
  group_by(genus) %>%
  summarise(genomes=n()) %>%
  arrange(-genomes) %>%
  filter(genus != "g__") %>% 
  pull(genus) %>% 
  head(n=100)
```

## New taxa discovery

```{r genome_taxonomy_stats_newgenomes, warning=FALSE, comments="", message=FALSE}
#species
genome_metadata %>% 
  mutate(novelty = if_else(species == "s__", "new", "known")) %>% 
  group_by(novelty) %>% 
    summarise(
      total_genomes = n(),
      percentage = (total_genomes / nrow(genome_metadata)) * 100
    ) %>% 
  tt()

#genus
genome_metadata %>% 
  mutate(novelty = if_else(genus == "g__", "new", "known")) %>% 
  group_by(novelty) %>% 
    summarise(
      total_genomes = n(),
      percentage = (total_genomes / nrow(genome_metadata)) * 100
    ) %>% 
  tt()

#order
genome_metadata %>% 
  mutate(novelty = if_else(order == "o__", "new", "known")) %>% 
  group_by(novelty) %>% 
    summarise(
      total_genomes = n(),
      percentage = (total_genomes / nrow(genome_metadata)) * 100
    ) %>% 
  tt()

```

```{r genome_taxonomy_stats_newgenomes_host, warning=FALSE, comments="", message=FALSE}
#species
genome_metadata %>% 
  left_join(assembly_metadata,by="assembly_id") %>% 
  left_join(summary_table,by="rawdata_id") %>% 
  mutate(novelty = if_else(species == "s__", "new", "known")) %>%
  group_by(host_species) %>%
  summarise(
    total_genomes = n(),
    new_genomes = sum(novelty == "new"),
    percentage_new = (new_genomes / total_genomes) * 100
  ) %>%
  arrange(desc(percentage_new)) %>% 
  tt()
```

## Microbial phylogenies

### Bacteria

```{r genome_tree_bacteria, warning=FALSE, comments="", message=FALSE}

gtdb_tree <- read_tree("data/bac120_r214.tree")

gtdb_metadata <- read_tsv("data/bac120_taxonomy_r214.tsv.gz", col_names = FALSE) %>% 
  rename(genome=1,classification=2) %>% 
  separate(classification, c("domain","phylum","class","order","family","genus","species"),  sep =";")

represented_genera <- gtdb_metadata %>% 
  filter(genus %in% genus_host$genus) %>% 
  filter(genome %in% gtdb_tree$tip.label) %>%
  group_by(genus) %>%
  slice_sample(n = 1) %>%
  ungroup()
  
ehi_mags_tree <- keep.tip(gtdb_tree, tip=represented_genera$genome)

ehi_mags_tree$tip.label <- represented_genera %>%
  filter(genome %in% ehi_mags_tree$tip.label) %>% # Keep only rows corresponding to tree tip labels
  arrange(match(genome, ehi_mags_tree$tip.label)) %>% # Match order of genomes to tree tip labels
  pull(genus)
```

```{r genome_tree_bacteria_plot, warning=FALSE, comments="", message=FALSE}
genus_metadata <- genome_metadata %>% 
  filter(phylum %in% selected_phyla) %>% 
  group_by(genus,phylum) %>% 
  summarise(count=n()) %>% 
  filter(genus %in% ehi_mags_tree$tip.label)

phylum_metadata <- genus_metadata %>%
  group_by(phylum) %>%
  summarise(node = MRCA(ehi_mags_tree, genus)) %>%
  arrange(node) %>% 
  filter(node>length(ehi_mags_tree$tip.label)) %>% 
  mutate(group = row_number()) 
 
ehi_mags_tree <- groupClade(ehi_mags_tree, phylum_metadata$node)

# Replace group numbers to phylum names
current_groups <- attributes(ehi_mags_tree)$group 
group_to_phylum <- setNames(phylum_metadata$phylum, phylum_metadata$group)
group_to_phylum <- c(`0` = NA,group_to_phylum)

new_groups <- sapply(current_groups, function(g) {
  if (g %in% phylum_metadata$group) {
    return(group_to_phylum[g])  # Map numeric group to phylum name
  } else {
    return(NA)  # Keep other nodes NA
  }
}) %>% unname()
attributes(ehi_mags_tree)$group <- new_groups

#Plot tree
ggtree(ehi_mags_tree, layout="equal_angle", aes(color = group)) %<+% genus_metadata +
  geom_tippoint(aes(size = count, alpha=0.5)) +
  scale_size_continuous(range = c(1, 10)) +
  scale_color_manual(values = phylum_colors, na.value = "black") +
  theme_void()

```

```{r genome_tree_archaea, warning=FALSE, comments="", message=FALSE}

gtdb_tree <- read_tree("data/ar53_r214.tree")

gtdb_metadata <- read_tsv("data/ar53_taxonomy_r214.tsv.gz", col_names = FALSE) %>% 
  rename(genome=1,classification=2) %>% 
  separate(classification, c("domain","phylum","class","order","family","genus","species"),  sep =";")

represented_genera <- gtdb_metadata %>% 
  filter(genus %in% genus_host$genus) %>% 
  filter(genome %in% gtdb_tree$tip.label) %>%
  group_by(genus) %>%
  slice_sample(n = 1) %>%
  ungroup()
  
ehi_mags_tree <- keep.tip(gtdb_tree, tip=represented_genera$genome)

ehi_mags_tree$tip.label <- represented_genera %>%
  filter(genome %in% ehi_mags_tree$tip.label) %>% # Keep only rows corresponding to tree tip labels
  arrange(match(genome, ehi_mags_tree$tip.label)) %>% # Match order of genomes to tree tip labels
  pull(genus)
```

```{r genome_tree_archaea_plot, warning=FALSE, comments="", message=FALSE}
genus_metadata <- genome_metadata %>% 
  filter(phylum %in% selected_phyla) %>% 
  group_by(genus,phylum) %>% 
  summarise(count=n()) %>% 
  filter(genus %in% ehi_mags_tree$tip.label)

phylum_metadata <- genus_metadata %>%
  group_by(phylum) %>%
  summarise(node = MRCA(ehi_mags_tree, genus)) %>%
  arrange(node) %>% 
  filter(node>length(ehi_mags_tree$tip.label)) %>% 
  mutate(group = row_number()) 
 
ehi_mags_tree <- groupClade(ehi_mags_tree, phylum_metadata$node)

# Replace group numbers to phylum names
current_groups <- attributes(ehi_mags_tree)$group 
group_to_phylum <- setNames(phylum_metadata$phylum, phylum_metadata$group)
group_to_phylum <- c(`0` = NA,group_to_phylum)

new_groups <- sapply(current_groups, function(g) {
  if (g %in% phylum_metadata$group) {
    return(group_to_phylum[g])  # Map numeric group to phylum name
  } else {
    return(NA)  # Keep other nodes NA
  }
}) %>% unname()
attributes(ehi_mags_tree)$group <- new_groups

#Plot tree
ggtree(ehi_mags_tree, layout="equal_angle", aes(color = group)) %<+% genus_metadata +
  geom_tippoint(aes(size = count, alpha=0.5)) +
  scale_size_continuous(range = c(1, 10)) +
  scale_color_manual(values = c("#c9c9c9","#8a8a8a","#454545"), na.value = "black") +
  theme_void()

```
## Genus - host heatmap

```{r genus_host_relationship, warning=FALSE, comments="", message=FALSE}
genus_host <- genome_metadata %>% 
  left_join(assembly_metadata,by="assembly_id") %>% 
  left_join(summary_table,by="rawdata_id") %>% 
  select(phylum,genus,host_species) %>% 
  filter(genus != "g__") %>% 
  group_by(phylum,genus,host_species) %>% 
  summarise(genomes=n()) %>% 
  arrange(-genomes) %>% 
  ungroup() 

genus_host_spread <- genus_host %>% 
    group_by(phylum,genus) %>% 
    summarise(host_species=n()) %>% 
    arrange(-host_species)

```


```{r genome_tree_plot, warning=FALSE, comments="", message=FALSE, fig.width=10, fig.height=18}
genus_host2 <- genus_host %>% 
  select(-phylum) %>%
  pivot_wider(names_from = host_species, values_from = genomes) %>% 
  column_to_rownames(var="genus")

genus_host %>% 
  left_join(host_metadata,by="host_species") %>% 
  mutate(genus=factor(genus,levels=ehi_mags_tree$tip.label)) %>% 
  ggplot(aes(x=host_species,y=genus,fill=genomes)) + 
    facet_nested(. ~ host_class, space="free",scales="free") +
    geom_tile() + 
    scale_fill_gradient(low = "grey", high = "#2a94d1", na.value="white") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genus_host, by=join_by(phylum == phylum)) %>%
    select(genus,phylum) %>% 
    unique %>% 
    mutate(genus=factor(genus,levels=ehi_mags_tree$tip.label)) %>% 
    mutate(xaxis=NA) %>% 
    ggplot(aes(x=xaxis,y=genus,fill=phylum)) +
    geom_tile() +
    scale_fill_manual(values=phylum_colors)
```
